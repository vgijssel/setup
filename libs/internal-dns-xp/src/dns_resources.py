"""DNS resources generation using Python (for testing and composition)."""

import secrets
from typing import Any, Dict, List


def create_dns_deployment(props: Dict[str, Any]) -> Dict[str, Any]:
    """Create DNS deployment with external-dns and PowerDNS containers."""
    domain = props.get("domain", "internal.local")
    external_dns_version = props.get("external_dns_version", "v0.14.0")
    powerdns_version = props.get("powerdns_version", "powerdns/pdns-auth-50")
    sync_interval = props.get("sync_interval", "30s")

    return {
        "apiVersion": "apps/v1",
        "kind": "Deployment",
        "metadata": {
            "name": "internal-dns-deployment",
            "labels": {"app": "internal-dns"},
        },
        "spec": {
            "replicas": 1,
            "selector": {"matchLabels": {"app": "internal-dns"}},
            "template": {
                "metadata": {"labels": {"app": "internal-dns"}},
                "spec": {
                    "containers": [
                        {
                            "name": "external-dns",
                            "image": f"registry.k8s.io/external-dns/external-dns:{external_dns_version}",
                            "args": [
                                "--source=ingress",
                                "--provider=pdns",
                                "--pdns-server=http://localhost:8081",
                                "--pdns-api-key=$(PDNS_API_KEY)",
                                f"--domain-filter={domain}",
                                f"--interval={sync_interval}",
                            ],
                            "env": [
                                {
                                    "name": "PDNS_API_KEY",
                                    "valueFrom": {
                                        "secretKeyRef": {
                                            "name": "powerdns-api-key",
                                            "key": "api-key",
                                        }
                                    },
                                }
                            ],
                        },
                        {
                            "name": "powerdns",
                            "image": powerdns_version,
                            "ports": [
                                {"containerPort": 53, "name": "dns", "protocol": "UDP"},
                                {
                                    "containerPort": 53,
                                    "name": "dns-tcp",
                                    "protocol": "TCP",
                                },
                                {"containerPort": 8081, "name": "api"},
                            ],
                            "volumeMounts": [
                                {"name": "dns-data", "mountPath": "/var/lib/powerdns"},
                                {"name": "config", "mountPath": "/etc/powerdns"},
                            ],
                        },
                    ],
                    "volumes": [
                        {
                            "name": "dns-data",
                            "persistentVolumeClaim": {"claimName": "dns-storage"},
                        },
                        {
                            "name": "config",
                            "configMap": {"name": "powerdns-config"},
                        },
                    ],
                },
            },
        },
    }


def create_dns_service(props: Dict[str, Any]) -> Dict[str, Any]:
    """Create DNS service."""
    return {
        "apiVersion": "v1",
        "kind": "Service",
        "metadata": {
            "name": "internal-dns-service",
            "labels": {"app": "internal-dns"},
        },
        "spec": {
            "type": "ClusterIP",
            "selector": {"app": "internal-dns"},
            "ports": [
                {"name": "dns-udp", "port": 53, "protocol": "UDP"},
                {"name": "dns-tcp", "port": 53, "protocol": "TCP"},
                {"name": "api", "port": 8081, "protocol": "TCP"},
            ],
        },
    }


def create_dns_pvc(props: Dict[str, Any]) -> Dict[str, Any]:
    """Create PersistentVolumeClaim for DNS storage."""
    storage_size = props.get("storage_size", "1Gi")

    return {
        "apiVersion": "v1",
        "kind": "PersistentVolumeClaim",
        "metadata": {
            "name": "dns-storage",
            "labels": {"app": "internal-dns"},
        },
        "spec": {
            "accessModes": ["ReadWriteOnce"],
            "resources": {"requests": {"storage": storage_size}},
        },
    }


def create_powerdns_config(props: Dict[str, Any]) -> Dict[str, Any]:
    """Create PowerDNS configuration ConfigMap."""
    pdns_conf = """launch=gsqlite3
gsqlite3-database=/var/lib/powerdns/pdns.sqlite3
webserver=yes
webserver-address=0.0.0.0
webserver-port=8081
api=yes
local-address=0.0.0.0
local-port=53"""

    return {
        "apiVersion": "v1",
        "kind": "ConfigMap",
        "metadata": {
            "name": "powerdns-config",
            "labels": {"app": "internal-dns"},
        },
        "data": {"pdns.conf": pdns_conf},
    }


def create_api_secret(props: Dict[str, Any]) -> Dict[str, Any]:
    """Create PowerDNS API key secret."""
    api_key = secrets.token_urlsafe(32)

    return {
        "apiVersion": "v1",
        "kind": "Secret",
        "metadata": {
            "name": "powerdns-api-key",
            "labels": {"app": "internal-dns"},
        },
        "type": "Opaque",
        "stringData": {"api-key": api_key},
    }


def create_all_dns_resources(props: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Create all DNS resources."""
    return [
        create_dns_deployment(props),
        create_dns_service(props),
        create_dns_pvc(props),
        create_powerdns_config(props),
        create_api_secret(props),
    ]
