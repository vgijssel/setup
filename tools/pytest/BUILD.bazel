load("@rules_python//python:pip.bzl", "compile_pip_requirements")

exports_files([
    "pytest_wrapper.py",
])

compile_pip_requirements(
    name = "requirements",
    extra_args = ["--allow-unsafe"],
    requirements_in = "requirements.in",
    requirements_txt = "requirements.lock",
)
