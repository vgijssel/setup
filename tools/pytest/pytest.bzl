load("@rules_python//python:defs.bzl", "py_test")
load("@pytest//:requirements.bzl", "requirement")

def py_pytest_test(name, srcs, deps = [], args = [], **kwargs):
    py_test(
        name = name,
        srcs = [
            "//tools/pytest:pytest_wrapper.py",
        ] + srcs,
        main = "//tools/pytest:pytest_wrapper.py",
        args = [
            "-ra",
            "-vv",
        ] + args + ["$(location :%s)" % x for x in srcs],
        python_version = "PY3",
        srcs_version = "PY3",
        deps = deps + [
            requirement("pytest"),
            requirement("pytest-cov"),
        ],
        **kwargs
    )
