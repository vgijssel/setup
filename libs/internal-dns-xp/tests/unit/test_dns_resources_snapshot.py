"""Snapshot test for all DNS resources."""

import json
import sys
from pathlib import Path

import pytest

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "src"))

from dns_resources import create_all_dns_resources  # noqa: E402


def test_dns_resources_snapshot(snapshot):
    """Test all DNS resources match snapshot."""
    props = {
        "domain": "internal.test.com",
        "storage_size": "1Gi",
        "external_dns_version": "v0.14.0",
        "powerdns_version": "powerdns/pdns-auth-50",
        "sync_interval": "30s",
    }

    resources = create_all_dns_resources(props)

    # Sort resources for consistent snapshot
    sorted_resources = sorted(
        resources, key=lambda r: (r["kind"], r["metadata"]["name"])
    )

    # Normalize random API key for snapshot comparison
    for resource in sorted_resources:
        if resource["kind"] == "Secret" and "stringData" in resource:
            if "api-key" in resource["stringData"]:
                resource["stringData"]["api-key"] = "NORMALIZED_API_KEY"

    assert snapshot == json.dumps(sorted_resources, indent=2, sort_keys=True)


def test_all_resources_created():
    """Test all required resources are created."""
    resources = create_all_dns_resources({"domain": "test.local"})

    kinds = {r["kind"] for r in resources}

    assert "Deployment" in kinds
    assert "Service" in kinds
    assert "PersistentVolumeClaim" in kinds
    assert "ConfigMap" in kinds
    assert "Secret" in kinds


def test_resources_have_labels():
    """Test all resources have app label."""
    resources = create_all_dns_resources({"domain": "test.local"})

    for resource in resources:
        assert "app" in resource["metadata"].get("labels", {})
        assert resource["metadata"]["labels"]["app"] == "internal-dns"


def test_resource_count():
    """Test correct number of resources are created."""
    resources = create_all_dns_resources({"domain": "test.local"})

    assert len(resources) == 5  # Deployment, Service, PVC, ConfigMap, Secret
