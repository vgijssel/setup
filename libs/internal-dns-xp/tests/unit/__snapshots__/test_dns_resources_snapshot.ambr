# serializer version: 1
# name: test_dns_resources_snapshot
  '''
  [
    {
      "apiVersion": "v1",
      "data": {
        "pdns.conf": "launch=gsqlite3\ngsqlite3-database=/var/lib/powerdns/pdns.sqlite3\nwebserver=yes\nwebserver-address=0.0.0.0\nwebserver-port=8081\napi=yes\nlocal-address=0.0.0.0\nlocal-port=53"
      },
      "kind": "ConfigMap",
      "metadata": {
        "labels": {
          "app": "internal-dns"
        },
        "name": "powerdns-config"
      }
    },
    {
      "apiVersion": "apps/v1",
      "kind": "Deployment",
      "metadata": {
        "labels": {
          "app": "internal-dns"
        },
        "name": "internal-dns-deployment"
      },
      "spec": {
        "replicas": 1,
        "selector": {
          "matchLabels": {
            "app": "internal-dns"
          }
        },
        "template": {
          "metadata": {
            "labels": {
              "app": "internal-dns"
            }
          },
          "spec": {
            "containers": [
              {
                "args": [
                  "--source=ingress",
                  "--provider=pdns",
                  "--pdns-server=http://localhost:8081",
                  "--pdns-api-key=$(PDNS_API_KEY)",
                  "--domain-filter=internal.test.com",
                  "--interval=30s"
                ],
                "env": [
                  {
                    "name": "PDNS_API_KEY",
                    "valueFrom": {
                      "secretKeyRef": {
                        "key": "api-key",
                        "name": "powerdns-api-key"
                      }
                    }
                  }
                ],
                "image": "registry.k8s.io/external-dns/external-dns:v0.14.0",
                "name": "external-dns"
              },
              {
                "image": "powerdns/pdns-auth-50",
                "name": "powerdns",
                "ports": [
                  {
                    "containerPort": 53,
                    "name": "dns",
                    "protocol": "UDP"
                  },
                  {
                    "containerPort": 53,
                    "name": "dns-tcp",
                    "protocol": "TCP"
                  },
                  {
                    "containerPort": 8081,
                    "name": "api"
                  }
                ],
                "volumeMounts": [
                  {
                    "mountPath": "/var/lib/powerdns",
                    "name": "dns-data"
                  },
                  {
                    "mountPath": "/etc/powerdns",
                    "name": "config"
                  }
                ]
              }
            ],
            "volumes": [
              {
                "name": "dns-data",
                "persistentVolumeClaim": {
                  "claimName": "dns-storage"
                }
              },
              {
                "configMap": {
                  "name": "powerdns-config"
                },
                "name": "config"
              }
            ]
          }
        }
      }
    },
    {
      "apiVersion": "v1",
      "kind": "PersistentVolumeClaim",
      "metadata": {
        "labels": {
          "app": "internal-dns"
        },
        "name": "dns-storage"
      },
      "spec": {
        "accessModes": [
          "ReadWriteOnce"
        ],
        "resources": {
          "requests": {
            "storage": "1Gi"
          }
        }
      }
    },
    {
      "apiVersion": "v1",
      "kind": "Secret",
      "metadata": {
        "labels": {
          "app": "internal-dns"
        },
        "name": "powerdns-api-key"
      },
      "stringData": {
        "api-key": "NORMALIZED_API_KEY"
      },
      "type": "Opaque"
    },
    {
      "apiVersion": "v1",
      "kind": "Service",
      "metadata": {
        "labels": {
          "app": "internal-dns"
        },
        "name": "internal-dns-service"
      },
      "spec": {
        "ports": [
          {
            "name": "dns-udp",
            "port": 53,
            "protocol": "UDP"
          },
          {
            "name": "dns-tcp",
            "port": 53,
            "protocol": "TCP"
          },
          {
            "name": "api",
            "port": 8081,
            "protocol": "TCP"
          }
        ],
        "selector": {
          "app": "internal-dns"
        },
        "type": "ClusterIP"
      }
    }
  ]
  '''
# ---
